src/protocols/MeterFile.cpp

```diff
@@ -169,21 +180,33 @@ ssize_t MeterFile::read(std::vector<Reading> &rds, size_t n) {
 	// wait for file change via inotify
 	const int EVENTSIZE = sizeof(struct inotify_event) + NAME_MAX + 1;
 	if (_notify_fd!=-1){
-	  // read all events from fd:
-	  char buf[EVENTSIZE *5];
-	  ssize_t len;
-	  int nr_events;
-	  do{
-            nr_events = 0;
-            len = ::read(_notify_fd, buf, sizeof(buf));
-            const struct inotify_event *event = (struct inotify_event *)(&buf[0]);
-            for (char *ptr = buf; ptr < buf + len;
-		 ptr += sizeof(struct inotify_event) + event->len) {
-	      ++nr_events;
-	      event = (const struct inotify_event *) ptr;
-	      print(log_debug, "got inotify_event %x", "file", event->mask);
-            }
-	  } while(len>0 && nr_events>=5); // if 5 events received there might be some more pending.
+		// read all events from fd:
+		char buf[EVENTSIZE *5];
+		ssize_t len;
+		int nr_events;
+		do{
+			nr_events = 0;
+			len = ::read(_notify_fd, buf, sizeof(buf));	// read will block until inotify event occurs
+			const struct inotify_event *event = (struct inotify_event *)(&buf[0]);
+			for (char *ptr = buf; ptr < buf + len; ptr += sizeof(struct inotify_event) + event->len) {
+				++nr_events;
+				event = (const struct inotify_event *) ptr;
+				print(log_debug, "got inotify_event %x", "file", event->mask);
+				if (event->mask & (IN_DELETE_SELF | IN_MOVE_SELF)) {
+					// File has been moved or deleted, therefore new inotify watch needed
+					if (_notify_fd != -1) {
+						(void)::close(_notify_fd);
+						if (inotify_add_watch(_notify_fd, path(), IN_CLOSE_WRITE | IN_DELETE_SELF | IN_MOVE_SELF) < 0) {
+							// Error: unable to add inotify watch, fall back to interval mechanism
+							print(log_alert, "inotify_add_watch(%s): %s", name().c_str(), path(), strerror(errno));
+							(void)::close(_notify_fd);
+							_notify_fd = -1;
+							_interval=1;	// assume interval length of 1 sec
+						}
+					}
+				}
+			}
+		} while (len>0 && nr_events>=5); // if 5 events received there might be some more pending.
```

Comment:
this might still block if exactly 5 (or more) events occurred. There might be no more pending.

Reply:
Hi Matthias,
what is your concrete suggestion? You are right that if there are >=5 inotify events, then only the first 5 get processed, and the next read() will get blocked until a new event comes in.
But what is the likelihood in "normal" operations that vzlogger gets swamped by so many writes to the watched file?
One fix could be to increase the event buffer from 5 to something larger.
I am running the modified code, so far no problems whatsoever.
-holger

Reply:
I'd suggest to do a:
int nr_events = 0;
do{
int totalPending = 0;
 if (nr_events) {
// no blocking expected on 2nd call

(void)ioctl(_notify_fd, FIONREAD, &totalPending); // we ignore retail and rely on totalPending changed only on success
}
if (nr_events==0 || totalPending>0)
len = ::read(...);
else len = 0;
nr_events = 0;
...